src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -164,177 +146,77 @@ private static CopyColumnsTransform Create(IHostEnvironment env, ModelLoadContex
+            ctx.Writer.Write(ColumnPairs.Length);
+            foreach (var column in ColumnPairs)
             {
-                ctx.SaveNonEmptyString(column.Name);
-                ctx.SaveNonEmptyString(column.Source);
+                ctx.SaveNonEmptyString(column.output);
+                ctx.SaveNonEmptyString(column.input);
             }
         }
 
-        private RowToRowMapperTransform CreateRowToRowMapper(IDataView input)
-        {
-            var mapper = new CopyColumnsRowMapper(_host, input.Schema, _columns);
-            return new RowToRowMapperTransform(_host, input, mapper);
-        }
-
-        public IDataView Transform(IDataView input)
-        {
-            return CreateRowToRowMapper(input);
-        }
-
-        public bool IsRowToRowMapper => true;
-
-        public IRowToRowMapper GetRowToRowMapper(ISchema inputSchema)
-        {
-            _host.CheckValue(inputSchema, nameof(inputSchema));
-            return CreateRowToRowMapper(new EmptyDataView(_host, inputSchema));
-        }
-    }
-
-    internal sealed class CopyColumnsRowMapper : IRowMapper
-    {
-        private readonly ISchema _schema;
-        private readonly Dictionary<int, int> _colNewToOldMapping;
-        private readonly (string Source, string Name)[] _columns;
-        private readonly IHost _host;
-        public const string LoaderSignature = "CopyColumnsRowMapper";
+        protected override IRowMapper MakeRowMapper(ISchema inputSchema)
+            => new Mapper(this, inputSchema, ColumnPairs);
 
-        private static VersionInfo GetVersionInfo()
+        private sealed class Mapper : MapperBase
         {
-            return new VersionInfo(
-                modelSignature: "COPYROWM",
-                verWrittenCur: 0x00010001, // Initial
-                verReadableCur: 0x00010001,
-                verWeCanReadBack: 0x00010001,
-                loaderSignature: LoaderSignature,
-                loaderAssemblyName: typeof(CopyColumnsRowMapper).Assembly.FullName);
-        }
+            private readonly ISchema _schema;
+            private readonly (string Source, string Name)[] _columns;
 
-        // Factory method for SignatureLoadRowMapper.
-        private static CopyColumnsRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, ISchema schema)
-        {
-            Contracts.CheckValue(env, nameof(env));
-            env.CheckValue(ctx, nameof(ctx));
-            ctx.CheckAtModel(GetVersionInfo());
-
-            // *** Binary format ***
-            // int: number of added columns
-            // for each added column
-            //   string: output column name
-            //   string: input column name
-
-            var length = ctx.Reader.ReadInt32();
-            var columns = new (string Source, string Name)[length];
-            for (int i = 0; i < length; i++)
+            public Mapper(CopyColumnsTransform parent, ISchema inputSchema, (string Source, string Name)[] columns)
```

Comment:
>public [](start = 12, length = 6)

would internal suffice?

Reply:
actually it should be private

---
In reply to: [222532234](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222532234) [](ancestors = 222532234)

Reply:
since it's a nested class which get created by higher in hierarchy, it can easily be private

---
In reply to: [222532234](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222532234) [](ancestors = 222532234)

Reply:
at this point I'm curious how many wrong comments can I leave here. :)
Technically it's a private class, so no one would see it any way....

---
In reply to: [222532234](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222532234) [](ancestors = 222532234)

Reply:
Mapper is a private class and a subclass to CopyColumnsTransform. I went ahead and make the constructor internal. @ivmatan@microsoft.com let me know if you have any additional feedback.

---
In reply to: [222533129](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222533129) [](ancestors = 222533129,222532234)

Reply:
I don't have any other feedback. So feel free to push your changes into PR, if that's the reason why you haven't done it yet :)

---
In reply to: [222782703](https://github.com/dotnet/machinelearning/pull/1141#discussion_r222782703) [](ancestors = 222782703,222533129,222532234)